common/flatpak-utils.c

```diff
@@ -5791,160 +5791,246 @@ get_write_progress (guint outstanding_writes)
-          else
-            new_progress = 10 + (87 * (double) fetched) / requested;
-        }
+      goto out;
     }
-  else if (outstanding_extra_data && downloading_extra_data)
+
+  if (outstanding_metadata_fetches > 0)
     {
-      guint64 elapsed_time =
-        (current_time - ostree_async_progress_get_uint64 (progress, "start-time-extra-data")) / G_USEC_PER_SEC;
-      guint64 bytes_sec = (elapsed_time > 0) ? transferred_extra_data_bytes / elapsed_time : 0;
-      g_autofree char *formatted_bytes_transferred =
-        g_format_size_full (transferred_extra_data_bytes, 0);
-      g_autofree char *formatted_bytes_sec = NULL;
+      /* At this point we don't really know how much data there is, so we have to make a guess.
+       * Since its really hard to figure out early how much data there is we report 1% until
+       * all objects are scanned. */
 
-      if (!bytes_sec) // Ignore first second
-        formatted_bytes_sec = g_strdup ("-");
-      else
-        formatted_bytes_sec = g_format_size (bytes_sec);
+      estimating = TRUE;
 
-      g_string_append_printf (buf, "Downloading extra data: %u%% (%" G_GUINT64_FORMAT "/%" G_GUINT64_FORMAT ") %s/s %s",
-                              (guint) ((((double) transferred_extra_data_bytes) / total_extra_data_bytes) * 100),
-                              transferred_extra_data_bytes, total_extra_data_bytes, formatted_bytes_sec, formatted_bytes_transferred);
+      g_string_append_printf (buf, "Receiving metadata objects: %u/(estimating) %s/s %s",
+                              metadata_fetched, formatted_bytes_sec, formatted_bytes_transferred);
 
-      /* Once we reach here, at least 55% of the data was fetched */
-      new_progress = 55 + (guint) (45 * ((double) transferred_extra_data_bytes) / total_extra_data_bytes);
+      /* Go up to 10% until the metadata is all fetched */
+      new_progress = get_metadata_progress (metadata_fetched, outstanding_metadata_fetches);
     }
-  else if (outstanding_writes)
+  else
     {
-      g_string_append_printf (buf, "Writing objects: %u", outstanding_writes);
+      /* This is the very specific case where we're downloading static deltas
+       * and nothing else. We can use bytes_transferred in this case, instead
+       * of fetched_delta_part_size.
+       *
+       * The reason for that is that fetched_delta_part_size is only updated
+       * after a delta object is fully downloaded, causing jumps in the progress.
+       */
+      if (outstanding_delta_fetches == outstanding_fetches)
+        total_transferred = bytes_transferred;
+
+      /* The download progress goes up to 97% */
+      new_progress = 5 + ((total_transferred / (gdouble) total) * 92);
 
-      /* Writing the objects advances 3% of progress */
-      new_progress = outstanding_extra_data ? 52 : 97;
+      /* And the writing of the objects adds 3% to the progress */
       new_progress += get_write_progress (outstanding_writes);
```

Comment:
This seems to do 3/outstanding_writes. Will that not do a division by zero when outstanding_writes is 0?